Add DNAME Support

rtokarek-fastly opened the pull request:
The Dyn API and library(python module) this provider uses supports DNAME records, but the provider did not. 

This change adds support for DNAME records and works around an issue with python 3.10 support. 

Dyn is being end-of-lifed in May 2023 which makes updates seem less important, but this might help some other people using OctoDNS to migrate off of Dyn DNS services. 

Support was added mostly by copying CNAME support except as DNAME. That will definitely probably work out fine.

rtokarek-fastly committed:
Add DNAME record support

We have a couple domains that use it and adding support should help anybody else importing data from Dyn in preparation for the impending May 2023 end-of-life of the service.

rtokarek-fastly committed:
Added tests for DNAME

Basically duplicates of CNAME tests

rtokarek-fastly committed:
Fixed Dynamic DNAME, deprecated python feature, and lint

The dyn libs we import here try to import Iterable from collections.

That's been moved into collections.abc (Abstract Base Class) so we're aliasing it here.

Suppress some linter errors about imports not being at the top of the file. They are functionally, but not literally. :eye-roll:

rtokarek-fastly committed:
Fix python 3.10 issue with collections and lint errors

python 3.10 deprecated collections.Iterable and moved it into collections.abc.Iterable. Import the latter and add an alias so the dyn modules we import that use the old pattern don't cause us problems here.

Suppress linter complaint about module imports not being at the literal top of the file. They're functionally at the top.

ross commented on octodns_dyn/__init__.py:
```diff
@@ -2,21 +2,27 @@
+# dyn libs needs the folowing alias as it's trying to use a
+# deprecated feature that worked up to python 3.9, but doesn't
+# in 3.10
+collections.Iterable = collections.abc.Iterable
```
Would it be possible to include everything you can normally w/o the `noqa` bits and then do the stuff that causes the need for it as the last bit so that only those few lines require it.

My guess/hope is that this collections line and some/all of the `from dyn*` bits a few lines down could all be moved after the other imports so that only the `from dyn*` lines need the marker.

rtokarek-fastly replied:
Yep, I've reordered them. The git diff is a little wonky looking, but it all seems to lint/test now with fewer # noqa: E402 lines.

ross reviewed (commented):
Overall seems good. Question for discussion inline.

rtokarek-fastly committed:
Reordered `imports`s to avoid suppressing E402 lint errors

Moved the various imports to avoid needing to suppress so many E402 lint errors (imports not at the top of the file) except for the specific dyn imports that require the collections.Iterable alias in python 3.10+ (works in python 3.3+, but wasn't a required workaround until 3.10). )
